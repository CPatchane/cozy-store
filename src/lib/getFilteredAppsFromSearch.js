import 'url-search-params-polyfill'

const getDoctypesList = (permsObj = {}) => {
  const doctypes = []
  for (let p in permsObj) {
    if (permsObj[p].type) doctypes.push(permsObj[p].type)
  }
  return doctypes
}

const getFilteredAppsFromSearch = (apps = [], search = '') => {
  let filteredApps = apps
  /* global URLSearchParams */
  const params = new URLSearchParams(search)

  // filter by type
  const typeParam = params.get('type')
  if (typeParam) filteredApps = filteredApps.filter(a => a.type === typeParam)

  // filer by category
  const categoryParam = params.get('category')
  if (categoryParam) filteredApps = filteredApps.filter(a => {
    if (!Array.isArray(a.categories)) return false
    return a.categories.includes(categoryParam)
  })

  // filter by tag
  const tagParam = params.get('tag')
  if (tagParam) filteredApps = filteredApps.filter(a => {
    if (!Array.isArray(a.tags)) return false
    return a.tags.includes(tagParam)
  })

  // filter by doctype
  const doctypeParam = params.get('doctype')
  if (doctypeParam) {
    filteredApps = filteredApps.filter(a => {
      if (!a.permissions) return false
      return getDoctypesList(a.permissions).includes(doctypeParam)
    })
  }

  return filteredApps
}

export default getFilteredAppsFromSearch
