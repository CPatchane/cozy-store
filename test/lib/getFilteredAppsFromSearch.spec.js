'use strict'

/* eslint-env jest */

import getFilteredAppsFromSearch from 'lib/getFilteredAppsFromSearch'
import mockApps from '../ducks/apps/_mockApps'

describe('getFilteredAppsFromSearch library', () => {
  it('should filter correctly on type', () => {
    expect(getFilteredAppsFromSearch(mockApps, '?type=webapp')).toMatchSnapshot()
  })

  it('should filter correctly on doctype', () => {
    expect(getFilteredAppsFromSearch(mockApps, '?doctype=io.mock.doctype2')).toMatchSnapshot()
  })

  it('should filter correctly type AND doctype', () => {
    expect(getFilteredAppsFromSearch(mockApps, '?type=konnector&doctype=io.mock.doctype')).toMatchSnapshot()
  })
})
