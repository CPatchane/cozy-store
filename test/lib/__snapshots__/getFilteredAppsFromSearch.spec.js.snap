// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFilteredAppsFromSearch library should filter correctly on doctype 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
]
`;

exports[`getFilteredAppsFromSearch library should filter correctly on type 1`] = `
Array [
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Collect",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://collect.cozy.mock/",
    "slug": "collect",
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0",
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "others",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "locales": Object {
      "name": "For Dev Only",
    },
    "name": "DevOnly",
    "slug": "devonly",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [],
      "dev": Array [
        "1.0.0-betaojdkehy989ekhflldh",
      ],
      "stable": Array [],
    },
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "installed": true,
    "name": "Drive",
    "name_prefix": "Cozy",
    "permissions": Object {
      "mock2": Object {
        "type": "io.mock.doctype2",
      },
    },
    "related": "http://drive.cozy.mock/",
    "slug": "drive",
    "type": "webapp",
    "uninstallable": false,
    "version": "3.0.0-beta89bnhj3993",
  },
  Object {
    "categories": Array [
      "cozy",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "editor": "Cozy",
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "name": "Photos",
    "name_prefix": "Cozy",
    "slug": "photos",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "3.0.0",
      ],
      "dev": Array [
        "3.0.0",
      ],
      "stable": Array [
        "3.0.0",
      ],
    },
  },
  Object {
    "categories": Array [
      "partners",
    ],
    "icon": "<svg></svg>",
    "isInRegistry": true,
    "name": "Tasky",
    "slug": "tasky",
    "type": "webapp",
    "uninstallable": true,
    "versions": Object {
      "beta": Array [
        "1.0.0",
      ],
      "dev": Array [
        "1.0.0",
      ],
      "stable": Array [
        "1.0.0",
      ],
    },
  },
]
`;

exports[`getFilteredAppsFromSearch library should filter correctly type AND doctype 1`] = `
Array [
  Object {
    "categories": Array [
      "transport",
    ],
    "developer": Object {
      "name": "Cozy",
    },
    "icon": "<svg></svg>",
    "installed": true,
    "isInRegistry": true,
    "name": "Trinlane",
    "permissions": Object {
      "mock": Object {
        "type": "io.mock.doctype",
      },
    },
    "slug": "konnector-trinlane",
    "type": "konnector",
    "uninstallable": true,
    "version": "0.1.0",
    "versions": Object {
      "beta": Array [
        "0.1.0",
      ],
      "dev": Array [
        "0.1.0",
      ],
      "stable": Array [
        "0.1.0",
      ],
    },
  },
]
`;
